packages/ember-application/tests/system/visit_test.js

```diff
@@ -90,49 +421,456 @@ if (isEnabled('ember-application-visit')) {
+      } else {
+        return instance.container.lookup(fullName);
       }
+    }
 
-      ok(lookup('-view-registry:main')['my-cool-app'] instanceof View, 'view was registered on the instance\'s view registry');
-      ok(lookup('-view-registry:main')['child-view'] instanceof View, 'child view was registered on the instance\'s view registry');
+    let a = run(function() {
+      return App.visit('/a', { browser: false, render: false }).then((instance) => {
+        QUnit.start();
+        assert.ok(instance instanceof ApplicationInstance, 'promise is resolved with an ApplicationInstance');
 
-      instance.destroy();
-    }, function(error) {
-      QUnit.start();
-      assert.ok(false, 'The visit() promise was rejected: ' + error);
+        assert.strictEqual(Ember.$('#qunit-fixture').children().length, 0, 'there are no elements in the fixture element');
+        assert.strictEqual(xFooInstances, 0, 'did not create any x-foo components');
+
+        let viewRegistry = lookup(instance, '-view-registry:main');
+        assert.strictEqual(Object.keys(viewRegistry).length, 0, 'did not create any views');
+
+        let networkService = lookup(instance, 'service:network');
+        assert.deepEqual(networkService.get('requests'), ['/a', '/b', '/c']);
+        QUnit.stop();
+      });
+    });
+
+    let b = run(function() {
+      return App.visit('/b', { browser: false, render: false }).then((instance) => {
+        QUnit.start();
+        assert.ok(instance instanceof ApplicationInstance, 'promise is resolved with an ApplicationInstance');
+
+        assert.strictEqual(Ember.$('#qunit-fixture').children().length, 0, 'there are no elements in the fixture element');
+        assert.strictEqual(xFooInstances, 0, 'did not create any x-foo components');
+
+        let viewRegistry = lookup(instance, '-view-registry:main');
+        assert.strictEqual(Object.keys(viewRegistry).length, 0, 'did not create any views');
+
+        let networkService = lookup(instance, 'service:network');
+        assert.deepEqual(networkService.get('requests'), ['/b', '/c']);
+        QUnit.stop();
+      });
+    });
+
+    let c = run(function() {
+      return App.visit('/c', { browser: false, render: false }).then((instance) => {
+        QUnit.start();
+        assert.ok(instance instanceof ApplicationInstance, 'promise is resolved with an ApplicationInstance');
+
+        assert.strictEqual(Ember.$('#qunit-fixture').children().length, 0, 'there are no elements in the fixture element');
+        assert.strictEqual(xFooInstances, 0, 'did not create any x-foo components');
+
+        let viewRegistry = lookup(instance, '-view-registry:main');
+        assert.strictEqual(Object.keys(viewRegistry).length, 0, 'did not create any views');
+
+        let networkService = lookup(instance, 'service:network');
+        assert.deepEqual(networkService.get('requests'), ['/c']);
+        QUnit.stop();
+      });
     });
+
+    let d = run(function() {
```

Comment:
I don't disagree. I admit that after working on this PR, whatever mental model I thought had about the runloop has been completely destroyed :P

I think I mostly just inherited these patterns from copying the existing tests. But in this particular case I believe removing it would fail the test because the first `visit` here calls `boot`, which has a `run.once("actions", ...);` line somewhere along the way.